src/modules/fw_pos_control/FixedwingPositionControl.cpp

```diff
@@ -756,28 +756,38 @@ FixedwingPositionControl::set_control_mode_current(const hrt_abstime &now)
 			_control_mode_current = FW_POSCTRL_MODE_AUTO;
 		}
 
+		if (_control_mode_current == FW_POSCTRL_MODE_AUTO && _vehicle_status.is_vtol && _vehicle_status.in_transition_mode) {
+			const float ground_speed = sqrt(_local_pos.vx * _local_pos.vx + _local_pos.vy * _local_pos.vy);
+			const float windspeed = _wind_vel.norm();
+
+			if (ground_speed < 2.f * _local_pos.evh && (!_wind_valid || windspeed < 2.f * _local_pos.evh)) {
+				_control_mode_current = FW_POSCTRL_MODE_AUTO_ALTITUDE_STRAIGHT;
+
+			}
+		}
+
```

Comment:
On where to apply it, my thinking was that this method (although i agree it is not as clean as it could be) is the closest to a "mode executor" which is the way forward. So we have one single place which decides what is appropriate control law ans then the control_auto_xxx (sort of the"modes") can only do their thing instead of having to do some mitigations when they are active, but can't actually work reliably, due to wrong conditions. Also this way, at least the logic is only in a single place instead of all the different control_auto_xx functions.

Reply:
> This condition would be for the case, that we don't move with respect to ground, and there is no wind, which means, we don't move at all! There it also does not make much sense to already use npfg, because in this instance it can't do anything anyways.

NPFG handles low airspeed with continuous ramp down to zero acceleration command. low ground speed + low (valid) wind speed = low airspeed. Note bearing feasibility is zeroed when airspeed is low, and each lateral acceleration function has airspeed multiplied in the output.

https://github.com/PX4/PX4-Autopilot/blob/0f260fc26c7bcfff3abb362e2306ac2c09c289c4/src/lib/npfg/npfg.cpp#L405-L467

however, wouldn't hurt to check a realistic condition and see if there are still some accel commands high enough that we dont like and want to add extra protection.

> On where to apply it, my thinking was that this method (although i agree it is not as clean as it could be) is the closest to a "mode executor" which is the way forward. So we have one single place which decides what is appropriate control law ans then the control_auto_xxx (sort of the"modes") can only do their thing instead of having to do some mitigations when they are active, but can't actually work reliably, due to wrong conditions. Also this way, at least the logic is only in a single place instead of all the different control_auto_xx functions.

That's also fair. This really only affects VTOLs in transition and auto mode. I just worry about hard switching logic with modes.. where these conditions may switch back and forth. Additionally, altitude control within the mode doesnt need to be hindered by degraded lateral conditions.

Maybe we need third opinion. As my opinion on this particular PR is getting less strong either way. (though for sure want to know how jumpy setpoints are handled)